src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,760 @@
+        let client_id = AtomicU64::new(0);
+        let inner_connection = RpcClientConnectionInner::new(
+            stream,
+            timeout_options.clone(),
+            client_id.load(std::sync::atomic::Ordering::Acquire),
+        );
+
+        Self {
+            timeout_options,
+            inner_connection: Arc::new(inner_connection),
+            client_id,
+        }
+    }
+
+    /// Start client receiver
+    pub fn start_recv(&self) {
+        // Create a receiver loop
+        let inner_connection_clone = Arc::clone(&self.inner_connection);
+        tokio::spawn(async move {
+            inner_connection_clone.recv_loop().await;
+        });
+    }
+
+    /// Send a request to the server.
+    /// Try to send data to channel, if the channel is full, return an error.
+    /// Contains the rezquest header and body.
+    /// WARN: this function does not support concurrent call
+    pub async fn send_request(&self, req: P) -> Result<(), RpcError> {
+        self.inner_connection
+            .send_packet(req)
+            .await
+            .map_err(|err| RpcError::InternalError(err.to_string()))
+    }
+
+    /// Manually send ping by the send request
+    /// The inner can not start two loop, because the stream is unsafe
+    /// we need to start the loop in the client or higher level manually
+    /// WARN: this function does not support concurrent call
+    pub async fn ping(&self) -> Result<(), RpcError> {
+        self.inner_connection.ping().await
+    }
+}
+
+/// `RecvLenFuture` TCP Stream reader future for receive data
+/// We need to keep block the stream until the data is received
+/// or the stream is closed or client is dropped
+/// The future will be used to receive the response from the server
+struct RecvLenFuture<'a, P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    /// The inner connection for the client.
+    client: &'a RpcClientConnectionInner<P>,
+    /// The length of the data to receive.
+    len: u64,
+    /// The buffer to store the received data.
+    req_buffer: &'a mut BytesMut,
+    /// The receiver for the closed signal.
+    is_closed: flume::Receiver<()>,
+}
+
+impl<'a, P> RecvLenFuture<'a, P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    /// Create a new receive length future.
+    fn new(client: &'a RpcClientConnectionInner<P>, len: u64, req_buffer: &'a mut BytesMut, is_closed: flume::Receiver<()>) -> Self {
+        Self {
+            client,
+            len,
+            req_buffer,
+            is_closed,
+        }
+    }
+}
+
+impl<P> Future for RecvLenFuture<'_, P>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+{
+    type Output = Result<(), RpcError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        // Recv closed data from the client channel
+        if this.is_closed.try_recv().is_ok() {
```

Comment:
You should poll the is_closed channel,  using try_recv() will not be notified to wake up

Reply:
Ok, this struct will use `recv_async` future and use `poll` to reach channel status.